src/engine/effects/engineeffectsdelay.cpp

```diff
@@ -0,0 +1,87 @@
+#include "engine/effects/engineeffectsdelay.h"
+
+#include "util/rampingvalue.h"
+#include "util/sample.h"
+
+namespace {
+// See enginedelay.cpp
+const int kiMaxDelay = static_cast<int>(0.508 *
+        mixxx::audio::SampleRate::kValueMax * mixxx::kEngineChannelCount);
+} // anonymous namespace
+
+EngineEffectsDelay::EngineEffectsDelay()
+        : m_currentDelaySamples(0),
+          m_prevDelaySamples(0),
+          m_delayBufferWritePos(0) {
+    m_pDelayBuffer = SampleUtil::alloc(kiMaxDelay);
+}
+
+void EngineEffectsDelay::process(CSAMPLE* M_RESTRICT pInOut,
+        const int iBufferSize) {
+    if (m_prevDelaySamples == 0 && m_currentDelaySamples == 0) {
+        return;
+    }
+
+    int delaySourcePos =
+            (m_delayBufferWritePos + kiMaxDelay - m_currentDelaySamples) %
+            kiMaxDelay;
```

Comment:
I agree, that from a mathematical point of view you're right. Anyway, from the C++ point of view, the modulo operator for negative values (for example, x%y, where x is a negative value) produces negative values (but in math result value is positive). The addition of delay buffer size is used to ensure positive values, otherwise, the calculation fails on the `"delaySourcePos >= 0"` debug assert.

Reply:
Yes, of course. I think I will create a new PR and document the same tricky part for `EngineDelay` and `EngineFilterDelay` because all these calculations are the same. As you wrote, it could be cool to solve this problem differently, but for now, I can't think of the more optimal solution.